x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateFormatter.java

```diff
@@ -0,0 +1,188 @@
+        );
+
+        Map<String, DateFormatter> formatterMap = new LinkedHashMap<>();
+        for (DateFormatter dateFormatter : formatters) {
+            formatterMap.put(dateFormatter.pattern, dateFormatter);
+        }
+        FORMATTER_MAP = formatterMap;
+    }
+
+    private static String ordinalSuffixForDayOfTheMonth(int i) {
+        if (i == 1 || i == 21 || i == 31) {
+            return "st";
+        } else if (i == 2 || i == 22) {
+            return "nd";
+        } else if (i == 3 || i == 23) {
+            return "rd";
+        } else {
+            return "th";
+        }
+
+    }
+
+    private String pattern;
+    private Function<TemporalAccessor, String> javaFormat;
+    private Function<String, String> additionalMapper;
+
+    public static Function<TemporalAccessor, String> ofPattern(String dateFormatPattern) {
+        if (Strings.isEmpty(dateFormatPattern)) {
+            return timestamp -> "";
+        }
+        List<DateFormatter> dateFormatters = parsePattern(dateFormatPattern);
+        return timestamp -> dateFormatters.stream().map(p -> p.format(timestamp)).collect(Collectors.joining());
+    }
+
+    private String format(TemporalAccessor timestamp) {
+        String formatted = this.javaFormat.apply(timestamp);
+        return additionalMapper == null ? formatted : this.additionalMapper.apply(formatted);
+    }
+
+    private static List<DateFormatter> parsePattern(String dateFormatPattern) {
+        LinkedList<DateFormatter> formatters = new LinkedList<>();
+        DateFormatter dateFormatter;
+
+        while (dateFormatPattern.length() > 1) {
+            String potentialPattern = dateFormatPattern.substring(0, 2);
+            dateFormatter = FORMATTER_MAP.get(potentialPattern);
+
+            if (dateFormatter != null) {
+                dateFormatPattern = dateFormatPattern.substring(2);
+            } else if (potentialPattern.startsWith("%")) {
+                dateFormatter = literal(dateFormatPattern.substring(1, 2));
+                dateFormatPattern = dateFormatPattern.substring(2);
+            } else if (potentialPattern.endsWith("%")) {
+                dateFormatter = literal(dateFormatPattern.substring(0, 1));
+                dateFormatPattern = dateFormatPattern.substring(1);
+            } else {
+                dateFormatter = literal(dateFormatPattern.substring(0, 2));
+                dateFormatPattern = dateFormatPattern.substring(2);
+            }
+
+            formatters.addLast(dateFormatter);
+
+        }
+        if (dateFormatPattern.length() == 1) {
+            dateFormatter = literal(dateFormatPattern.substring(0, 1));
+            formatters.addLast(dateFormatter);
+        }
+
+        return formatters;
+    }
+
+    private static DateFormatter literal(String literal) {
+        DateFormatter dateFormatter = new DateFormatter();
+        dateFormatter.javaFormat = timestamp -> literal;
+        return dateFormatter;
+    }
+
+    public static class Builder {
```

Comment:
This class can also be made entirely `private` (including all its methods).

Reply:
ok